converters/toZigbee.js

```diff
@@ -1429,6 +1429,80 @@ const converters = {
             await entity.read('closuresDoorLock', ['lockState']);
         },
     },
+    pincode_lock: {
+        key: ['pin_code'],
+        convertSet: async (entity, key, value, meta) => {
+            const userId = value.userId;
+            const pinCode = value.pinCode;
+            if ( isNaN(userId) ) {
+                throw new Error('userId must be numbers');
+            }
+            if (converters.pincode_lock.userIdInRange(meta, userId)) {
+                throw new Error('userId must be in range for device');
+            }
+            if (pinCode === undefined || pinCode === null) {
+                await entity.command(
+                    'closuresDoorLock',
+                    'clearPinCode',
+                    {
+                        'userid': userId,
+                    },
+                    getOptions(meta.mapped),
+                );
+            } else {
+                if (isNaN(pinCode)) {
+                    throw new Error('pinCode must be a number or pinCode');
+                }
+                await entity.command(
+                    'closuresDoorLock',
+                    'setPinCode',
+                    {
+                        'userid': userId,
+                        'userstatus': 1,
+                        'usertype': 0,
+                        'pincodevalue': pinCode.toString(),
+                    },
+                    getOptions(meta.mapped),
+                );
+            }
+            return {readAfterWriteTime: 200};
+        },
+        convertGet: async (entity, key, meta) => {
+            const userId = meta && meta.message && meta.message.pin_code ? meta.message.pin_code.userId : undefined;
```

Comment:
In case no userID is provided, can we just get all user codes like you do in `onReconnect` now? This would also make `onReconnect` obsolete.